WMF Framework/Event Platform Client/EventPlatformClientProtocols.swift

```diff
@@ -0,0 +1,101 @@
+/*
+ * Event Platform Client (EPC)
+ *
+ * DESCRIPTION
+ *     Collects events in an input buffer, adds some metadata, places them in an
+ *     ouput buffer where they are periodically bursted to a remote endpoint via
+ *     HTTP POST.
+ *
+ *     Designed for use with Wikipedia iOS application producing events to the
+ *     EventGate intake service.
+ *
+ * AUTHORS
+ *     Mikhail Popov <mpopov@wikimedia.org>
+ *     Jason Linehan <jlinehan@wikimedia.org>
+ *
+ * LICENSE NOTICE
+ *     Copyright 2019 Wikimedia Foundation
+ *
+ *     Redistribution and use in source and binary forms, with or without
+ *     modification, are permitted provided that the following conditions are
+ *     met:
+ *
+ *     1. Redistributions of source code must retain the above copyright notice,
+ *     this list of conditions and the following disclaimer.
+ *
+ *     2. Redistributions in binary form must reproduce the above copyright
+ *     notice, this list of conditions and the following disclaimer in the
+ *     documentation and/or other materials provided with the distribution.
+ *
+ *     THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+ *     IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO,
+ *     THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
+ *     PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR
+ *     CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
+ *     EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+ *     PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
+ *     PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF
+ *     LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+ *     NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+ *     SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+import Foundation
+
+/**
+ * Storage manager provides EPC with the ability to persist data and retrieve persisted data, asynchronously.
+ *
+ * The storage manager needs to be able to handle persistent storage of any type of value. When storing a
+ * session identifier, for example, it needs to store a `String`. When storing queued-up, to-be-sent events
+ * when entering background, it needs to store `[(String, String)]`. When storing raw logged events,
```

Comment:
Yep, although the only use right now is saving/loading the `input_buffer` (to be renamed to `inputBuffer`) of events which were logged before the stream configuration is available and need to be processed once the stream config is downloaded.

So the actual concrete scenario is an array of events as tuples: `[(stream: String, schema: String, data: [String: Any], domain: String?)]`

Reply:
Do you have recommendations for `data`'s type instead of a Dictionary of `Any`s? Because some event data fields might be strings, some might be numeric, some boolean, some dictionaries. JSON Schema (used by Event Platform for schema specification) supports objects in fields, so

```JSON
{
  "user": {
    "name": "Bearloga",
    "id": 150561
  },
  "session_length_s": 3600,
  "first_app_launch": false,
  "meta": {
     "stream": "test.sessions"
   },
  "$schema": "/analytics/test/session/1.0.0",
  "client_dt": "2020-07-21T09:54:01Z"
}
```

is an example of a POST body that the event ingestion service can handle (provided an appropriate schema)

So we want to be able to persist an event logged with
```swift
EPC.shared.log(
  stream: "test.sessions",
  schema: "/analytics/test/session/1.0.0",
  data: [
     "user": ["name": "Bearloga", "id": 150561],
     "session_length_s": 3600,
     "first_app_launch": false
  ]
)
```

Reply:
Sorry, JUST remembered your notes from later in the review

Reply:
I've renamed everything to use camel case instead of snake case and caps. I'm sorry if it was really annoying to see those naming conventions throughout the PR. When I upload the fixes it should be less annoying to review.